src/sntp.rs

```diff
@@ -122,23 +122,29 @@ pub struct EspSntp {
 
 impl EspSntp {
     pub fn new_default() -> Result<Self, EspError> {
-        Self::new(&Default::default())
+        Self::new(&Default::default(), None)
     }
 
-    pub fn new(conf: &SntpConf) -> Result<Self, EspError> {
+    pub fn new(
```

Comment:
* Rather than asking the user to deal with an unsafe "C" fn and with an unsafe `*mut timeval`, how about asking the user to provide a safe callback? As in e.g. `FnMut(core::time::Duration) + Send + 'static`? This would also be symmetric with what `SysTime` returns. Which is a `core::time::Duration`.
* You are introducing a backwards incompatible change. Shall we instead keep backwards compatibility and introduce a new `subscribe` method that takes the callback? As in: `pub fn subscribe<F: FnMut(core::time::Duration) + Send + 'static>(&mut self) -> EspSntpSubscription`? And then, dropping the `SntpSubscription would mean that the user is unsubscribing?

Reply:
Thanks for the advice, but I'm running into this - how do I get around the type conversion for `F`?

```rs
error[E0308]: mismatched types
   --> /home/mdesilva/esp/esp-idf-svc_fork/src/sntp.rs:172:53
    |
167 |     pub fn subscribe<F: FnMut(core::time::Duration) + Send + 'static>(
    |                      - this type parameter
...
172 |             sntp_set_time_sync_notification_cb(Some(callback));
    |                                                ---- ^^^^^^^^ expected fn pointer, found type parameter `F`
    |                                                |
    |                                                arguments to this enum variant are incorrect
    |
    = note:  expected fn pointer `unsafe extern "C" fn(*mut esp_idf_sys::timeval)`
            found type parameter `F`
```

A backwards incompatible change may be required (I need to test again), but you'll notice earlier `sntp_init()` is called after the callback is set.  I need to check if (1) we can set more than one callback, which relates to your other PR comment and (2) if these need to be set before the `sntp_init()` call.

Reply:
Hmm, for the type conversion I could do

```rs
#[allow(dead_code)]
pub struct Newtype<T>(pub T);

impl<T> Into<unsafe extern "C" fn(*mut esp_idf_sys::timeval)> for Newtype<T> {
    fn into(self) -> unsafe extern "C" fn(*mut esp_idf_sys::timeval) {
        todo!()
    }
}
```

with

```rs
    pub fn subscribe<F: FnMut(core::time::Duration) + Send + 'static>(
        &mut self,
        callback: F,
    ) -> EspSntpSubscription {
        unsafe {
            sntp_set_time_sync_notification_cb(Some(Newtype(callback).into()));
        }

        EspSntpSubscription {}
    }
    ```
    
    Let me know if this is an acceptable approach (not fully completed yet of course), thanks!